Yank/multistate/multistatereporter.py

```diff
@@ -700,62 +702,77 @@ def read_energies(self, iteration=slice(None)):
-        energy_thermodynamic_states = self._storage_analysis.variables['energies'][iteration, :, :]
+        energy_thermodynamic_states = np.array(self._storage_analysis.variables['energies'][iteration, :, :], np.float64)
+        energy_neighborhoods = np.array(self._storage_analysis.variables['neighborhoods'][iteration, :, :], 'i1')
         try:
-            energy_unsampled_states = self._storage_analysis.variables['unsampled_energies'][iteration, :, :]
+            energy_unsampled_states = np.array(self._storage_analysis.variables['unsampled_energies'][iteration, :, :], np.float64)
         except KeyError:
             # There are no unsampled thermodynamic states.
             unsampled_shape = energy_thermodynamic_states.shape[:-1] + (0,)
             energy_unsampled_states = np.zeros(unsampled_shape)
-        return energy_thermodynamic_states, energy_unsampled_states
+        return energy_thermodynamic_states, energy_neighborhoods, energy_unsampled_states
 
-    def write_energies(self, energy_thermodynamic_states, energy_unsampled_states, iteration):
+    def write_energies(self, energy_thermodynamic_states, energy_neighborhoods, energy_unsampled_states, iteration):
         """Store the energy matrix at the given iteration on the analysis file
 
         Parameters
         ----------
         energy_thermodynamic_states : n_replicas x n_states numpy.ndarray
             ``energy_thermodynamic_states[i][j]`` is the reduced potential computed at
             SamplerState ``sampler_states[i]`` and ThermodynamicState ``thermodynamic_states[j]``.
+        energy_neighborhoods : n_replicas x n_states numpy.ndarray
+            energy_neighborhoods[replica_index, state_index] is 1 if the energy was computed for this state,
+            0 otherwise
         energy_unsampled_states : n_replicas x n_unsampled_states numpy.ndarray
             ``energy_unsampled_states[i][j]`` is the reduced potential computed at SamplerState
             ``sampler_states[i]`` and ThermodynamicState ``unsampled_thermodynamic_states[j]``.
         iteration : int
             The iteration at which to store the data.
 
         """
-        # Initialize schema if needed.
-        if 'energies' not in self._storage_analysis.variables:
-            n_replicas, n_states = energy_thermodynamic_states.shape
+        n_replicas, n_states = energy_thermodynamic_states.shape
 
-            # Create dimensions if they weren't created by other functions.
-            self._ensure_dimension_exists('replica', n_replicas)
-            self._ensure_dimension_exists('state', n_states)
-
-            # Create variable for thermodynamic state energies with units and descriptions.
+        # Create dimensions and variables if they weren't created by other functions.
+        self._ensure_dimension_exists('replica', n_replicas)
+        self._ensure_dimension_exists('state', n_states)
+        if 'energies' not in self._storage_analysis.variables:
```

Comment:
The logic of lazy creation was changed so that things are created as they are needed.
One hope was that this would let us silently upgrade old-style netcdf files by adding dense neighborhoods if we try to resume with the post-SAMS code.